internal/regex/regex_onigmo.go

```diff
@@ -0,0 +1,34 @@
+// +build onigmo
+
+package regex
+
+import (
+	"runtime"
+
+	onigmo "github.com/linyows/go-onigmo"
+)
+
+type Onigmo struct {
+	re *onigmo.Regexp
+}
+
+func init() {
+	err := Register("onigmo", NewOnigmo)
+	if err != nil {
+		panic(err.Error())
+	}
+}
+
+func NewOnigmo(pattern string) (Matcher, error) {
+	re, err := onigmo.NewRegexp(pattern)
+	if err != nil {
+		return nil, err
+	}
+	runtime.SetFinalizer(re, (*onigmo.Regexp).Free)
```

Comment:
Finalizers are not guaranteed to be called. I don't think it's a proper mechanism to free the regexps.

Reply:
The safest refactoring (IMO) is to extend our matcher interface by adding close/free function. Or maybe we can sacrifice some performance and have only one Match function which will be registered and do all - compile/match/free.
WDYT?

Reply:
I think we can check in the usage place if the matcher implements `io.Closer` and then implement `Close() error` method only in the matchers that need some closing.

That way, current matchers do not need any refactoring and nothing _needs_ to change. Then, in the place we use the matcher we add the `io.Closer` spec and call it if needed.